### PR TITLE
Make SRAMs directly emit FIRRTL memories

### PR DESCRIPTION
**Pull request by debs-sifive**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)
- Backend code generation
- Internal or build-related (includes code refactoring/cleanup)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

`SRAM`s updated to directly emit FIRRTL memories (as opposed to creating `SyncReadMem`s.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/util/SRAM.scala
Comment (debs-sifive) [inline src/main/scala/chisel3/util/SRAM.scala:626]: @jackkoenig should these bundles live here discreetly? The file is getting kind of long, so I was thinking about separating these out into a different file.

File: core/src/main/scala/chisel3/internal/firrtl/IR.scala
Comment (debs-sifive) [inline core/src/main/scala/chisel3/internal/firrtl/IR.scala:291]: Thoughts on naming this `DefMem` and renaming `DefMem` to `CDefMem` for name alignment? Or not that important?

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (jackkoenig) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:10]: You can combine these two imports

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (jackkoenig) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:12]: You can merge this with the `import chisel3.util.SRAM` above

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (seldridge) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:53]: This make me so happy...

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (jackkoenig) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:41]: The default name can't be `_sram_MEM`. The reason this changed is that `val mem` in `SRAM.scala` is of type `SyncReadMem` which gets handled by the compiler plugin. You can do the naming yourself by wrapping the RHS of `val sram = ` with `autoNameRecursively` [[1]](https://github.com/chipsalliance/chisel/blob/9177535ff1be47ffd99034bf0154c0f1ec637419/core/src/main/scala/chisel3/internal/plugin/package.scala#L30) (make sure to leave a comment explaining why you're doing this).

File: src/main/scala/chisel3/util/SRAM.scala
Comment (jackkoenig) [reply src/main/scala/chisel3/util/SRAM.scala:626]: I think they're fine here, but leave a comment about why they're private and mention that the write and readwrite bundles are missing `mask` (and explain why).

File: src/main/scala/chisel3/util/SRAM.scala
Comment (jackkoenig) [inline src/main/scala/chisel3/util/SRAM.scala:546]:
```suggestion
    for((memWritePort, firrtlWritePort) <- _out.writePorts.zip(firrtlWritePorts)) {
      firrtlWritePort.addr := memWritePort.address
      firrtlWritePort.clk := Module.clock
      firrtlWritePort.data.asInstanceOf[Data] := memWritePort.data
      firrtlWritePort.en := memWritePort.enable
      assignMask(memWritePort.data, memWritePort.mask, firrtlWritePort.getRef, "mask")
    }
```
This is a not a strongly held position, but for simple foreaches like this, I prefer `for` style since it reduces a level of indention

File: src/main/scala/chisel3/util/SRAM.scala
Comment (jackkoenig) [inline src/main/scala/chisel3/util/SRAM.scala:507]:
```suggestion
    def nameAndSetRef(ports: Seq[Data], namePrefix: String): Seq[Name] = {
      ports.zipWithIndex.map { case (p, idx) =>
        val name = namePrefix + idx
        p.setRef(Slot(Node(mem), name))
        name        
      }
    }
    val firrtlReadPortNames: Seq[String] = nameAndSetRef(firrtlReadPorts, "R")
    val firrtlWritePortNames: Seq[String] = nameAndSetRef(firrtlWritePorts, "W")
    val firrtlReadwritePortNames: Seq[String] = nameAndSetRef(firrtlReadwritePorts, "RW")
```
Reduce, reuse, recycle

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (jackkoenig) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:142]: Good job writing good tests 🙂

File: core/src/main/scala/chisel3/internal/firrtl/IR.scala
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:291]: I don't think it's important, fine as is IMO

File: core/src/main/scala/chisel3/RawModule.scala
Comment (jackkoenig) [inline core/src/main/scala/chisel3/RawModule.scala:165]:
```suggestion
      id.forceName(default = "MEM", _namespace)
```
This is not related to my comment below because if you do that, we should never actually hit this default, but we should just default to `MEM` (like `MemBase` above)

**Review by jackkoenig — APPROVED** [review]
LGTM, mostly style nits, please fix the names of the memories before merging.

File: src/test/scala/chiselTests/util/SRAMSpec.scala
Comment (mwachs5) [inline src/test/scala/chiselTests/util/SRAMSpec.scala:189]: shoudl these tests use `:<>=` instead of `<>`?

File: src/main/scala/chisel3/util/SRAM.scala
Comment (debs-sifive) [reply src/main/scala/chisel3/util/SRAM.scala:546]: oh cool, that does look nice! I have never seen or used this pattern before :x

**Review by jackkoenig — APPROVED** [review]

**Review by seldridge — APPROVED** [review]

File: core/src/main/scala/chisel3/internal/firrtl/IR.scala
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:222]: Good eye, so recall (or learn very quickly) that `Arg` is the Chisel IR Expression. `Node` is the `Arg` that wraps a `HasId`. In SRAM, we manually build some nested Args to emit the connections to the mask fields of the ports because representing those as actual Chisel types is really hard. We needed to build some Slots that didn't have real Chisel Data backing them up (thus no `HasIds`). We figured it was pretty weird that `Slot` _required_ a `HasId` (whereas other `Args` like `Index` don't), so this change makes more sense anyway and let us do what we needed to do to emit the connections for the mask.

Comment (seldridge) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:222]: Follow up question: https://github.com/chipsalliance/chisel/pull/3955/files#diff-f01d2f59c3998b1bb4f17da0c509aa21205a6682b1cd42e274dca0fa92ebf33dR477 is preventing this alternative id-less `Slot` from showing up when doing a select query. This seems to make sense with your explanation.

Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:222]: Yeah that is the only place where this change had any effect. @debs-sifive maybe we should write a test on what happens if you do a Select query for connections on a Module with an SRAM. These connections to the real ports of the SRAM are iffy and that Select may let users get at them...

Comment (debs-sifive) [reply core/src/main/scala/chisel3/internal/firrtl/IR.scala:222]: I played around with this a bit, and we can indeed see some of the memory ports, but it's all output stuff:
```
List(PredicatedConnect(List(),MemWrapper.foo.readPorts[0].data: Wire[UInt<8>],MemWrapper.foo_sram.R0.data: [UInt<8>],false), PredicatedConnect(List(),MemWrapper.foo.readwritePorts[0].readData: Wire[UInt<8>],MemWrapper.foo_sram.RW0.rdata: [UInt<8>],false))
```

So it looks like we can't drive anything on the actual memory with Select. If we want Select functionality with SRAM, I say we do it in a separate PR :)

**Comment by sequencer** [comment]
Really appreciate. And we finally have this feature! I wanna kill all `Mem` usage for a long time. After this feature leading to chisel, maybe we can finally deprecate `Mem` API. 
Some questions but I don’t think it will block this PR:

- does it support `suggestName`?(some time sram isn’t sram, but RF). Originally it doesn’t because it warp `cmem`. 

- maybe in the future. It can be directly lowered to SRAM Intrinsic? Then we can simply add more interface to sram. For example extend MBIST interface.

- is possible to emit sram with Instance Select API for avoiding different lower pass for memory behavior models?

**Comment by debs-sifive** [comment]
@sequencer I haven't really worked with the Instance Select API, so not sure about that Q.

> does it support suggestName?(some time sram isn’t sram, but RF). Originally it doesn’t because it warp cmem.

It currently supports `suggestName` of the `SRAMInterface`. I believe we can support `suggestName` of the memory itself, since the memory target is a `HasID`. But it's not implemented in this iteration.

**Pull request merged**
